### ISSUE TITLE
Why is currying advantageous?

### ISSUE THREAD

**Issue by radiovisual** [learn-javascript-courses/javascript-questions#7]
I have just started to explore all the hype around functional programming, and writing "pure functions", so I am starting to see a hear a lot of new buzz words float around the javascript community. The one that currently has my attention is 'currying', which I understand is the process of breaking a function with multiple parameters into a set of multiple functions, each accepting only one parameter. 

I know that these concepts are second-nature to developers in functional languages, like Haskell, who's programming syntax force them to think in terms of currying and one-parameter functions, but I am yet to see the benefit of taking this:

    function multiply(x, y) { return x * y; }

and refactoring it to look like this:

    function curriedMultiply(x) {
      return function(y) { return x * y; }
    }

I hear such benefits like being able to "partially apply" functions, or making code that's more modular, testable, etc...But I can't see how returning a cascade of anonymous functions makes my code more testable, or modular...

I do get, however, that functional programming overall can make code far more readable, and easier to test and maintain, but the advantages of currying, and how it applies to functional programming are escaping me at the moment, because I feel I am not seeing the bigger picture. I have only found small examples like the one above, so I can't see the benefits yet, I only see a cascade of nested functions.

**So could you please present a real-world, practical example of why currying is advantageous?**

**Comment by ericelliott** [comment]
The first advantage is that it helps you remove state from your functions. James Coglan mentioned it in his talk, [Practical Functional Programming](https://www.youtube.com/watch?v=XcS-LdEBUkE).

The features James talks about in that talk depend on function type uniformity. Functions have types. For instance, a function that takes a function and an array and returns an array (like Function.prototype.map):

`fn, array -> array`

All the cool things he talks about rely on function type uniformity. For example he talks about how promises can help abstract your program logic's dependence on time. Promises do that because they **lift** whatever function you're calling so that the return type is always uniform: All functions that return promises return the same type, which means that you can use them in standardized ways. James talks about lots of ways that promises help.

In this case, promises are containers... by dealing with containers instead of the values inside of containers, you can create lots of generic functions that will work universally for anything that interacts with the container's interface.

Like promises, **curried functions** return containers that always return things that all share the same interface (functions are the containers in this case), and you just keep calling the function until something **other than** the function is returned.

In other words, a curried function is a function that lifts all of its parameters so that you can deal with those parameters in a standardized way.

For more on the importance of lifting functions, see also the "Function Polymorphism" section from Chapter 2 of [Programming JavaScript Applications](https://ericelliottjs.com/product/programming-javascript-applications-paper-ebook-bundle/) - Scroll down to the TOC, every section is linked to the free HTML version of my book.

You can read a lot more about these things here: [The Two Pillars of JavaScript Part 2](https://medium.com/javascript-scene/the-two-pillars-of-javascript-pt-2-functional-programming-a63aa53a41a4).

**Comment by radiovisual** [comment]
As always, you have a nice way of introducing complex topics in a manner that is easy-to-follow. Thanks! :+1:

**Comment by ericelliott** [comment]
:+1:

**Comment by ericelliott** [comment]
I updated this answer to clarify a few things that I left out before. Is it still clear?

**Comment by radiovisual** [comment]
Awesome! Thanks! Its still clear and super helpful! :computer: :bowtie:

**Comment by ericelliott** [comment]
\o/ :grinning:

**Comment by oscarms8** [comment]
ericelliott... Very well explained. I like your answer. Very good. Very good.

**Comment by ericelliott** [comment]
I turned this into a blog post:

[Curry or Partial Application?](https://medium.com/javascript-scene/curry-or-partial-application-8150044c78b8). Check it out.

**Comment by JoyalToTheWorld** [comment]
So let me get this straight:
 
```
const next = a(b);
doSomeStuff().then(() => next(c));
```

is WAY better than:

`doSomeStuff().then(() => a(b,c));`

**Comment by ericelliott** [comment]
@JoyalToTheWorld 

All curried functions are [higher order functions](https://medium.com/javascript-scene/higher-order-functions-composing-software-5365cf2cbe99).

Like all higher order functions, currying is about *abstraction.* It's commonly used to make function composition easier. During the process of abstraction, we're planning to support potentially more than one use-case, meaning that when we do `const next = a(b);` we probably don't know about `c` yet, or there may be potentially many values for `c`, in which case, yes, the curried version is clearly better, because the other version might imply code logic duplication, more code, or both.

**Comment by JoyalToTheWorld** [comment]
@ericelliott 

I understand the concepts of higher order functions and abstraction, however I am failing to see applicable use cases for this particular abstraction. 

I find the basic examples you provide have no foundation in reality and generally make it much more difficult to learn the concepts you are trying to teach. 

A solution in want of a problem is itself a problem. 
An abstraction that serves no useful purpose inherently makes things more complicated. 

Why wouldn't you know about `c`? And if you didn't know about `c` why are you calling `a` before you know about `c`? What do you lose by ensuring you have your necessary parameters before calling a function. Seems fairly straight forward to me, UNLESS you have a special situation, which is what we really wanted to talk about, but I can't seem to find any mention of this.

**Comment by ericelliott** [comment]
@JoyalToTheWorld you would have a good point, but dealing with asynchronous timing is extremely common in JavaScript. You may not know about `c` because it's a value you expect to get in the future (see promises, tasks, monads, etc...), but `a(b)` may still be a useful value that you can use in other situations, by applying potentially *many different values* for `c`.

If you really want to understand this, you should practice using `compose()` and `pipe()` to structure your application and build up larger functions from smaller building blocks. See ["What is Function Composition?"](https://medium.com/javascript-scene/master-the-javascript-interview-what-is-function-composition-20dfb109a1a0) for more details.

I've intentionally kept the examples here vague and over-simplified because digging too deeply into it would require an in-depth explanation of function composition, asynchronous programming, promises/futures, monads, and other category-theory stuff that is well out of scope and unnecessary to grasp the basic concept, which is simply:

**You can partially apply a function now and finish the application later at some unspecified time in the future.**

If you don't see it right now, you'll just have to trust me. If you don't trust me, why are you reading my stuff?
Sometimes too much detail is much worse than *vague hints* that a concept will come in handy later.

If you want the rest of the details, I've laid some of it out in a **17-part blog post series** called ["Composing Software"](https://medium.com/javascript-scene/composing-software-an-introduction-27b72500d6ea). There are more parts to come, and there's still a whole lot more to say about why deferring complete applications is very often useful.
